Refactor var expansion

rgrinberg opened the pull request:
This is a preparatory PR for making OSL allow forms such as `%{read:..}` work. There's no functional difference, but it does allow expanding values into OSL in a more consistent way.

emillon commented on src/super_context.ml:
```diff
@@ -552,40 +551,240 @@ module Scope_key = struct
+  | Infer
+  | Alias
+
+module Expander : sig
```
Is there a way to split that to its own file or does it depend too much on what's before in the file?

rgrinberg replied:
It depends on the `Super_context.t` itself. Once we have multi dir libs, we should be able to organize this better.

diml commented on src/ordered_set_lang.ml:
```diff
@@ -291,31 +297,40 @@ module Unexpanded = struct
+            match f fn with
+            | [x] -> Value.to_path ~dir x
+            | _ ->
+              Exn.code_error "Ordered_set_lang.Unexpanded.expand path"
```
I believe we can reach this case with `(:include %{cc})`, so it shouldn't be a code error

rgrinberg replied:
Ah ok, I see. Will do that.

diml commented on src/super_context.ml:
```diff
@@ -552,40 +551,240 @@ module Scope_key = struct
+
+  type expander
+    =  sctx
+    -> dir:Path.t
```
Why not pass all these arguments to `with_expander`? This should just be `String_with_vars.Var.t -> Syntax.Version.t -> Value.t list option` to match the signature of the function taken by `Action.Unexpanded.Partial_expand`.

rgrinberg replied:
Okay, I think that's good as well. Do we want to label those args btw? In pform, I chose the `~syntax_version` and `~pform` labels but I don't have a preference for which one is better. We should do it consistently everywhere however.

diml replied:
Not labeling them is fine given that they have different and specific types

rgrinberg commented:
@diml have another look. things are quite uniform now.

rgrinberg commented:
I'm merging because this is just a pre-req refactor PR for variables in OSL. I've addressed all the review comments, and I'm happy to address any other issues that might come up later.
